### PR TITLE
Revert "Revert D24923679: Fixed einsum compatibility/performance issues (#46398)"

### PR DESCRIPTION
**Pull request by heitorschueroff**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#49189 Revert "Revert D24923679: Fixed einsum compatibility/performance issues (#46398)"**

This reverts commit d307601365c3b848072b8b8381208aedc1a0aca5 and fixes the bug with diagonals and ellipsis combined.

Differential Revision: [D25540722](https://our.internmc.facebook.com/intern/diff/D25540722)

### CODE REVIEW

File: aten/src/ATen/native/Linear.cpp [hunk 1]
```diff
@@ -336,36 +336,36 @@ Tensor einsum(std::string equation, TensorList operands) {
...
+    const auto labels = op_labels[i];
+    const auto original_sizes = operand.sizes();
 
-    for (int label : labels) {
+    std::size_t j = 0;
+    for (auto label : labels) {
       if (label == ELLIPSIS) {
-        // Add missing dimensions under ellipsis
-        int64_t num_dim_diff =
-            ell_num_dim - (operand.dim() - labels.size() + 1);
```
Comment (heitorschueroff) [inline aten/src/ATen/native/Linear.cpp]: The bug was here. I was using the current operand size to compute how many dimensions were covered by the ellipsis. But if I happened to take a diagonal before the ellipsis the dimensions of operand would change and this value would be wrong.

File: aten/src/ATen/native/Linear.cpp [hunk 2]
```diff
@@ -336,36 +336,36 @@ Tensor einsum(std::string equation, TensorList operands) {
...
         }
         for (int64_t k = 0; k < ell_num_dim; ++k) {
           perm_shape[ell_index + k] = j++;
         }
       } else if (label_dim[label] != -1) {
         // Repeated label, take diagonal
-        int64_t dim = label_dim[label];
+        const auto dim = label_dim[label];
         TORCH_CHECK(
-            sizes[j] == sizes[dim],
```
Comment (heitorschueroff) [inline aten/src/ATen/native/Linear.cpp]: This was also wrong as sizes did not update as dimensions were added to operand.

File: test/test_linalg.py [hunk 3]
```diff
@@ -2696,6 +2697,59 @@ def check(equation, *operands):
         # with strided tensors
         check("bn,anm,bm->ba", l[:, ::2], w[:, ::2, ::2], r[:, ::2])
 
+    @dtypes(torch.double)
```
Comment (heitorschueroff) [inline test/test_linalg.py:2700]: I added a random test to hopefully catch anything missed in the manual test cases.

**Review by mruberry — APPROVED** [review]
Nice new randomized test.

File: test/test_linalg.py [hunk 4]
```diff
@@ -2620,6 +2620,202 @@ def test_old_matrix_rank(self, device, dtype):
             self.assertEqual(torch.matrix_rank(aaT, True), np.linalg.matrix_rank(aaT.cpu().numpy(), True))
             self.assertEqual(torch.matrix_rank(aaT, 0.01, True), np.linalg.matrix_rank(aaT.cpu().numpy(), 0.01, True))
 
+    @dtypes(torch.double)
```
Comment (IvanYashchuk) [inline test/test_linalg.py:2623]: Can we have complex dtype be tested here as well?
```suggestion
    @dtypes(torch.double, torch.cdouble)
```

**Pull request closed without merging**